### PR TITLE
Fix TestServer hang with duplex streaming requests

### PR DESCRIPTION
**Pull request by JamesNK**
#### Description

TestServer does not correctly execute bidirectional requests. The HttpClient from a TestServer request will hang reading the response in two situations:

1. The server writes no response content
2. The client stream is not completed.

This PR resolves both of these issues.

#### Customer Impact

Because of these issues, TestServer can't be used to host gRPC functional tests of bidirectional gRPC calls.

For gRPC functional testing customers will be forced to host an in-process Kestrel instance and call via TCP sockets. This could cause port exhaustion on some OS with a large number of tests.

#### Regression?

No, TestServer has never correctly supported bidirectional requests.
		
#### Risk

Low. TestServer is designed for unit tests, not production use. Note: TestServer ships in Microsoft.AspNetCore.TestHost, an out-of-sdk NuGet package: https://www.nuget.org/packages/Microsoft.AspNetCore.TestHost/

### CODE REVIEW

**Review by Tratcher — APPROVED** [review]

**Comment by thefringeninja** [comment]
Any chance of getting this into 3.1?

**Comment by anurse** [comment]
> Any chance of getting this into 3.1?

Not for 3.1.0, it's locked down tight. It could be considered for a patch to 3.1.1. @JamesNK is there any workaround for this issue?

**Comment by Tratcher** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=429434

**Comment by Tratcher** [comment]
Calling Response.Body.FlushAsync or Response.CompleteAsync should mitigate it.

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build.

**Comment by JamesNK** [comment]
Might be able to have a middleware in the test server instance that does nothing other than flush/complete. That wouldn’t be too intrusive.

Otherwise this could be a 3.1 patch.

**Comment by JamesNK** [comment]
> Calling Response.Body.FlushAsync or Response.CompleteAsync should mitigate it.

I've tested and this isn't enough. The response will be returned to the HttpClient, but the response stream won't close until the request stream is complete. That is because clean up will wait for the request to be completed before the response:

```cs
await CompleteRequestAsync();
await CompleteResponseAsync();
```

I think we'll need to do this as a 3.1 patch.

**Comment by JamesNK** [comment]
@anurse Added servicing label. Also update description with required template.

**Comment by anurse** [comment]
Do you want this to go in to 3.1 directly or backport from master? The PR is targeting master but you put servicing labels on it :).

If you want it to go to 3.1, retarget to `release/3.1` and put the label back on :). I'll take it to tactics.

**Comment by JamesNK** [comment]
I'd like to put it directly into 3.1 and then have it merged back to master. There is no urgency to get it into 5.0.

I'll retarget.

**Comment by anurse** [comment]
I want to have a chat with you offline about this one @JamesNK .

File: src/Hosting/TestHost/src/HttpContextBuilder.cs
Comment (halter73) [inline src/Hosting/TestHost/src/HttpContextBuilder.cs]: I think reordering CompleteResponseAsync and CompleteRequestAsync makes some sense since the client could hypothetically be awaiting a full response before completing its request.

Not awaiting the _sendRequestStreamTask worries me a little. What if it throws a non-OperationCanceledException and we don't observe it now? I'm not sure that logging is sufficient since the test might be over by the time any error is logged if CompleteRequestAsync doesn't await it.

Other than a situation like in the ClientStreaming_ResponseCompletesWithoutResponseBodyWrite test where the request stream task is intentionally held up until the end of the test, why would awaiting the _sendRequestStreamTask be problematic?

File: src/Hosting/TestHost/test/TestClientTests.cs
Comment (halter73) [inline src/Hosting/TestHost/test/TestClientTests.cs]: What exception does this throw? Kestrel has a 5-second drain timeout where the client can still send request body data after the response is completed, so this doesn't exactly match Kestrel's behavior.

Comment (Tratcher) [reply src/Hosting/TestHost/test/TestClientTests.cs]: TestServer does not try to preserve client observable behavior, only app visible behavior. At this point the app delegate has returned so there's no need to keep the request stream open.

Kestrel also has that drain timeout to avoid closing connections and there are no connections here to be concerned with.

File: src/Hosting/TestHost/src/HttpContextBuilder.cs
Comment (Tratcher) [reply src/Hosting/TestHost/src/HttpContextBuilder.cs]: > Other than a situation like in the ClientStreaming_ResponseCompletesWithoutResponseBodyWrite test where the request stream task is intentionally held up until the end of the test, why would awaiting the _sendRequestStreamTask be problematic?

A delayed request body is enough, it puts you into a deadlock condition.

File: src/Hosting/TestHost/test/TestClientTests.cs
Comment (JamesNK) [reply src/Hosting/TestHost/test/TestClientTests.cs]: > What exception does this throw?

I believe it is an IOException. The request stream throws it once the request is complete on the server.

**Review by halter73 — APPROVED** [review]

**Comment by anurse** [comment]
@Tratcher helped me clear this up.

* In 3.0, bidirectional requests worked
* We made changes in 3.1 that *regressed* this behavior
* This PR is to fix those changes.

Given that, this definitely is worth submitting for patch. Reapplied the labels.

**Comment by JamesNK** [comment]
No.

* 3.0, bidirectional requests hung.
* 3.1 added support for bidirectional requests. However it was discovered from users that the two scenarios listed here still hung.
* This PR fixes the two scenarios.

**Comment by mkArtakMSFT** [comment]
These were approved for 3.1.2, so updating the milestone accordingly /cc @jamshedd

**Comment by vivmishra** [comment]
Moved to Mar as per Tactics. Will need to be explicitly approved for Feb, if required.

**Pull request merged**
